tests/unit/plugins/inventory/test_opennebula.py

```diff
@@ -58,7 +58,21 @@ def get_vm_pool():
         'ETIME': 0,
         'GID': 132,
         'GNAME': 'CSApparelVDC',
-        'HISTORY_RECORDS': {},
+        'HISTORY_RECORDS': [
+            {
+                'HISTORY': OrderedDict({
+                    'OID': '42',
+                    'SEQ': '384',
+                    'HOSTNAME': 'sam-691-sam',
+                    'HID': '10',
+                    'CID': '0',
+                    'DS_ID': '100',
+                    'VM_MAD': 'kvm',
+                    'TM_MAD': '3par',
+                    'ACTION': '0'
+                }),
+            }
+        ],
```

Comment:
Maybe something like:
```suggestion
        'HISTORY_RECORDS': {
            'HISTORY' : [
                {
                    'OID': '42',
                    'SEQ': '384',
                    'HOSTNAME': 'sam-691-sam',
                    'HID': '10',
                    'CID': '0',
                    'DS_ID': '100',
                    'VM_MAD': 'kvm',
                    'TM_MAD': '3par',
                    'ACTION': '0',
                },
            ],
        },
```
By the way, this construct `OrderedDict({ ... })` that appears throughout this test code is likely redundant, they could all be replaced with the `dict` content `{...}`.

Reply:
Thanks @russoz. Since my `opennebula.py` inventory change does work when run against OpenNebula servers, I am reasonably sure that changing HISTORY_RECORDS back to a {} rather than a [] will break the test, which means it doesn't reflect the actual data structure that ON returns.

**If the code is working and the tests are currently passing, is this PR 'good enough' now?**

I'm quite burnt out on it and don't want to make myself dislike adding features as I'll be using Open Nebula for the next year plus!

Reply:
> Thanks @russoz. Since my `opennebula.py` inventory change does work when run against OpenNebula servers, I am reasonably sure that changing HISTORY_RECORDS back to a {} rather than a [] will break the test, which means it doesn't reflect the actual data structure that ON returns.
> 
> **If the code is working and the tests are currently passing, is this PR 'good enough' now?**

Well, the concern is that the piece of code that seems to parse that data goes like this:

```python
            if hasattr(vm.HISTORY_RECORDS, 'HISTORY'):
                if isinstance(vm.HISTORY_RECORDS.HISTORY, Sequence) and len(vm.HISTORY_RECORDS.HISTORY) > 0:
                    if hasattr(vm.HISTORY_RECORDS.HISTORY[-1], 'HOSTNAME'):
                        server['host'] = vm.HISTORY_RECORDS.HISTORY[-1].HOSTNAME
```

That seems to indicate that `HISTORY_RECORDS` has an **attribute** `HISTORY`, which in its turn is a **Sequence** (with at least one element), from which you select the last element, and the attribute `HOSTNAME` out of it. That looks not coherent with the statement that `HISTORY_RECORDS` is a list. Could you please explain why that difference?

People other than you (me, Felix, @feldsam and future maintainers of the collection) will need to provide support to this code for potentially a long time and we cannot explain it. Either we are missing some piece of the puzzle, or the test might doing the wrong thing.

So, answering your question, to be 'good enough' we need to understand what's happening. Hope you don't take this badly, but the long-term health of the code is more important than short-term results.

> I'm quite burnt out on it and don't want to make myself dislike adding features as I'll be using Open Nebula for the next year plus!

Sorry to hear that, hope you can find some time to decompress. There is no pressure from our side to solve this quickly, so take your time. 

All the best.

Reply:
Something like that should produce the data structure that the inventory plugin actually expects:
```python
class HistoryEntry(object):
    def __init__(self):
        self.SEQ: '384'
        self.HOSTNAME: 'sam-691-sam'
        self.HID: '10'
        self.CID: '0'
        self.DS_ID: '100'
        self.VM_MAD: 'kvm'
        self.TM_MAD: '3par'
        self.ACTION: '0'

class HistoryRecords(object):
    def __init__(self):
        self.HISTORY = [HistoryEntry()]

def get_vm_pool():
    ...
        'HISTORY_RECORDS': HistoryRecords(),
    ...
```

Reply:
As for the tests I like @felixfontein solution so we will see in a minute if it works :)